File: api/src/graphql/quiz_template.rs
use std::str::FromStr;

use juniper::{graphql_object, FieldResult, GraphQLInputObject, GraphQLObject, ID};
use mongodb::bson::oid::ObjectId;

use super::question::Question;

use crate::context::Context;
use crate::models;

struct QuizTemplate {
    id: ID,
    name: String,
}

impl QuizTemplate {
    fn from_model(model: &models::QuizTemplate) -> Self {
        println!("{:?}", model);
        Self {
            id: model.id.to_string().into(),
            name: model.name.clone(),
        }
    }
}

#[graphql_object]
#[graphql(context = Context)]
impl QuizTemplate {
    fn id(&self) -> ID {
        self.id.clone()
    }

    fn name(&self) -> &str {
        self.name.as_str()
    }

    async fn questions(&self, context: &Context) -> FieldResult<Vec<Question>> {
        let id = ObjectId::from_str(&self.id.to_string())?;

        let results = models::Question::get_by_quiz_template_id(&context.database, &id).await?;

        Ok(results.iter().map(Question::from_model).collect())
    }
}

#[derive(GraphQLInputObject)]
struct CreateQuizTemplate {
    name: String,
}

impl CreateQuizTemplate {
    fn to_model(&self) -> models::NewQuizTemplate {
        models::NewQuizTemplate {
            name: self.name.clone(),
        }
    }
}

#[derive(GraphQLObject)]
struct CreateQuizTemplateError {
    id: String,
}

#[derive(GraphQLInputObject)]
struct EditQuizTemplate {
    name: String,
}

impl EditQuizTemplate {
    fn to_model(&self) -> models::UpdateQuizTemplate {
        models::UpdateQuizTemplate {
            name: self.name.clone(),
        }
    }
}

pub struct QuizTemplateMutation;

#[graphql_object]
#[graphql(context = Context)]
impl QuizTemplateMutation {
    async fn create(&self, context: &Context, input: CreateQuizTemplate) -> FieldResult<QuizTemplate> {
        let input_model = input.to_model();

        let model = models::QuizTemplate::create(&context.database, &input_model).await?;

        Ok(QuizTemplate::from_model(&model))
    }

    async fn delete_by_id(&self, context: &Context, id: ID) -> FieldResult<bool> {
        let db_id = ObjectId::from_str(&id)?;

        let result = models::QuizTemplate::delete_by_id(&context.database, &db_id).await?;

        Ok(result)
    }

    async fn edit(
        &self,
        context: &Context,
        id: ID,
        input: EditQuizTemplate,
    ) -> FieldResult<QuizTemplate> {
        let input_model = input.to_model();

        let db_id = ObjectId::from_str(&id)?;

        let model = models::QuizTemplate::update(&context.database, &db_id, &input_model).await?;

        Ok(QuizTemplate::from_model(&model))
    }
}

pub struct QuizTemplateQuery;

#[graphql_object]
#[graphql(context = Context)]
impl QuizTemplateQuery {
    async fn all(&self, context: &Context) -> FieldResult<Vec<QuizTemplate>> {
        let models = models::QuizTemplate::get_all(&context.database).await?;

        Ok(models.iter().map(QuizTemplate::from_model).collect())
    }

    async fn by_id(&self, context: &Context, id: ID) -> FieldResult<Option<QuizTemplate>> {
        let db_id = ObjectId::from_str(&id)?;

        let model = models::QuizTemplate::get_by_id(&context.database, &db_id).await?;

        Ok(match model {
            Some(model) => Some(QuizTemplate::from_model(&model)),
            None => None,
        })
    }
}

File: api/app/resolvers/quiz_template.py
from bson import ObjectId
from typing import TYPE_CHECKING, Annotated
from strawberry import ID, Private, field, input, lazy, type

from ..context import get_context
from ..models import QuizTemplateModel, QuizTemplates

if TYPE_CHECKING:
    from .question import Question


def get_quiz_templates() -> QuizTemplates:
    return get_context().quiz_templates


@type
class QuizTemplate:
    id: ID
    name: str
    user_id: Private[ID]

    @staticmethod
    def from_model(model: QuizTemplateModel) -> "QuizTemplate":
        return QuizTemplate(
            id=ID(str(model["_id"])),
            name=model["name"],
            user_id=ID(str(model["user_id"])),
        )

    @field
    def questions(self) -> list[Annotated["Question", lazy(".question")]]:
        from .question import Question

        return list(
            map(
                Question.from_model,
                get_context().questions.get_by_quiz_template_id(ObjectId(str(self.id))),
            )
        )


@input
class CreateQuizTemplate:
    name: str
    user_id: ID


@input
class EditQuizTemplate(CreateQuizTemplate):
    id: ID


@type
class QuizTemplateMutation:
    @field
    def create(self, input: CreateQuizTemplate) -> QuizTemplate:
        return QuizTemplate.from_model(
            get_quiz_templates().create(
                {"name": input.name, "user_id": ObjectId(str(input.user_id))}
            )
        )

    @field
    def edit(self, input: EditQuizTemplate) -> QuizTemplate:
        return QuizTemplate.from_model(
            get_quiz_templates().edit_by_id(
                ObjectId(str(input.id)),
                {"name": input.name, "user_id": ObjectId(str(input.user_id))},
            )
        )

    @field
    def delete_by_id(self, id: ID) -> bool:
        return get_quiz_templates().delete_by_id(ObjectId(str(id)))


@type
class QuizTemplateQuery:
    @field
    def all(self) -> list[QuizTemplate]:
        return list(map(QuizTemplate.from_model, get_quiz_templates().all()))

    @field
    def by_id(self, id: ID) -> QuizTemplate:
        return QuizTemplate.from_model(
            get_quiz_templates().get_by_id(ObjectId(str(id)))
        )

    @field
    def by_user_id(self, user_id: ID) -> list[QuizTemplate]:
        return list(
            map(
                QuizTemplate.from_model,
                get_quiz_templates().get_by_user_id(ObjectId(str(user_id))),
            )
        )
